log-levels.md

```diff
@@ -0,0 +1,159 @@
+# Log Levels 
+
+This guide serves as guidelines for library authors with regards to what [SwiftLog](https://github.com/apple/swift-log) log levels are apropriate for use in libraries, and in what siuations to use what level.
+
+## Guidelines for Libraries
+
+SwiftLog defines the following 7 log levels via the [`Logger.Level` enum](https://apple.github.io/swift-log/docs/current/Logging/Structs/Logger/Level.html), ordered from least to most severe: 
+
+* `trace`
+* `debug`
+* `info`
+* `notice`
+* `warning`
+* `error`
+* `critical`
+
+Out of those, only levels _less severe than_ info (exclusive) are generally okay to be used by libraries.
+In the following section we'll explore how to use them in practice.
+
+### Recommended log levels
+
+It is always fine for a library to log at `trace` and `debug` levels, and these two should be the primary levels any library is logging at.
+
+`trace` is the finest log level, and end-users of a library will not usually use it unless debugging very specific issues. You should consider it as a way for library developers to "log everything we could possibly need to diagnose a hard to reproduce bug". It is expected to take a toll on the performance of a system, and developers can assume trace level logging will not be used in production deployments, unless enabled specifically to locate some specific issue.
+
+This is in contrast with `debug` which some users _may_ choose to run with enabled on their production systems.
+
+> Debug level logging should be not "too" noisy. Developers should assume some production deployments may need to (or want to) run with debug level logging enabled. 
+>
+> Debug level logging should not completely undermine the performance of a production system.
+
+As such, `debug` logging should not be overly noisy. It should provide a high value understanding of what is going on in the library for end users without diving deep into its internals. This is what `trace` is intended for.
```

Comment:
I think we should add (here or elsewhere) that you should try to make `debug` self-contained. What I mean with that you have enough information in just `debug` (without `trace`) to correlate two messages.

So for example if I see

```
debug: conection established [connection-id: 7]
debug: connection closed unexpectedly [error: foobar]
```

then I have a problem because I don't know if the second message is related to the first. So what I would suggest is (and what is done in AHC):
- make sure that there is _one_ ID (usually the request ID) that is logged in all messages related to that activity. (ie. in the above example, "connection closed unexpectedly" should also have `connection-id: 7`
- no need to log all IDs of everything every time, just the 'correlation ID' is enough. For example in AHC, we log a log of stuff in the 'starting request` message and in all follow-ups we just log the connection ID to be able to correlate it with the 'starting request' message that has all the other info (like the TCP connection etc).

Reply:
the reason of course is: If we need `trace` to make sense of `debug`, then `trace` == `debug` :P

Reply:
good hint, will add 👍

Reply:
Added a section on this, let me know what you think -- edits welcome :)